sample/bin/setup.js

```diff
@@ -0,0 +1,95 @@
+#!/usr/bin/env node
+
+const fs = require('fs');
+const rmrf = require('rmfr');
+const rmrfsync = require('rimraf').sync;
+const { main, MainError } = require('@effection/node');
+
+const { 
+  messages, yarn, TARGET_DIR, SOURCE_DIR, startScript
+} = require('./constants');
+const { formatErr, formatSuccess, spin } = require('./console-helpers');
+const { install } = require('./install');
+const { version } = require('../package.json');
+
+async function populate(message) {
+  if (fs.existsSync(TARGET_DIR)) {
+    throw new MainError({ 
+      message: `${formatErr('directory \'bigtest-sample\' already exists')}\n${messages.abort}`
+    });
+  } else {
+    fs.mkdirSync(TARGET_DIR);
+    const installer = {
+      name: 'bigtest-sample-temporary-package-name',
+      version: '0.0.0',
+      dependencies: {
+        'bigtest-sample': version
+      }
+    };
+    fs.writeFileSync(
+      `${TARGET_DIR}/package.json`,
+      JSON.stringify(installer, null, 2)
+    );
+    console.log(`\n${formatSuccess(message)}`);
+  };
+};
+
+function* migrate(messages) {
+  yield spin(messages.before, function*() {
+    const lockfile = yarn ? 'yarn.lock' : 'package-lock.json';
+    yield rmrf(`${TARGET_DIR}/${lockfile}`);
+    yield rmrf(`${TARGET_DIR}/package.json`);
+  
+    fs.readdirSync(SOURCE_DIR).filter(file => {
+      return file !== 'bin';
+    }).forEach(file => {
+      if(file==='package.json'){
+        const {
+          name, version, description, repository, author, license, 
+          main, scripts, devDependencies, eslintConfig, browserslist,
+          babel, jest
+        } = require(`${SOURCE_DIR}/${file}`);
+  
+        scripts.start = startScript;
+  
+        const pkgjson = {
+          name, version, private: true, description, repository, author, 
+          license, main, scripts, devDependencies, eslintConfig, browserslist,
+          babel, jest
+        };
+  
+        fs.writeFileSync(`${TARGET_DIR}/package.json`, JSON.stringify(pkgjson, null, 2));
+      } else {
+        fs.renameSync(`${SOURCE_DIR}/${file}`, `${TARGET_DIR}/${file}`);
+      };
+    });
+  
+    yield rmrf(`${TARGET_DIR}/node_modules/`);
+  });
+
+  console.log(formatSuccess(messages.after));
+};
+
+function* download(messages) {
+  yield spin(messages.before, install({ cwd: TARGET_DIR }));
+  console.log(formatSuccess(messages.after));
+};
+
+function* run() {
+  let rollback = true;
+  yield populate(messages.creating_dir);
```

Comment:
Note that you won't rollback if `populate()` fails or you halt here because you won't have entered the `try{}` is that intentional?

Reply:
> This is looking really good. How does it feel?

Very excited to finally learn how effection works. 👍 I don't mind doing a rewrite of this in Typescript but should we scope that to a separate PR or do that as the next step?

> Note that you won't rollback if populate() fails or you halt here because you won't have entered the try{} is that intentional?

populate() will throw an error if `sample-app` directory already exists. so i think forcing a rollback on that function might unintentionally delete someone's previous progress/experiments with bigtest. i could either keep it as is or throw in `rollback = false` inside the if block so that the entire function can move into the try/catch:
```
if (fs.existsSync(TARGET_DIR)) {
  rollback = false;
  throw new MainError({ 
    message: `${formatErr('directory \'bigtest-sample\' already exists')}\n${messages.abort}`
  });
}
```